core/src/main/java/ai/timefold/solver/core/impl/bavet/common/AbstractPrecomputeNode.java

```diff
@@ -0,0 +1,184 @@
+        }
+    }
+
+    @Override
+    public final void settle() {
+        if (!queuedRetractSet.isEmpty() || !queuedInsertSet.isEmpty()) {
+            invalidateCache();
+            queuedUpdateSet.removeAll(queuedRetractSet);
+            queuedUpdateSet.removeAll(queuedInsertSet);
+            // Do not remove queued retracts from inserts; if a fact property
+            // change, there will be both a retract and insert for that fact
+            queuedRetractSet.forEach(this::retractFromInnerNodeNetwork);
+            queuedInsertSet.forEach(this::insertIntoInnerNodeNetwork);
+            queuedRetractSet.clear();
+            queuedInsertSet.clear();
+
+            // settle the inner node network, so the inserts/retracts do not interfere
+            // with the recording of the first object's tuples
+            innerNodeNetwork.settle();
+            recalculateTuples();
+        }
+        for (var updatedObject : queuedUpdateSet) {
+            for (var updatedTuple : objectToOutputTuplesMap.get(updatedObject)) {
+                updateExisting(updatedObject, updatedTuple);
+            }
+        }
+        queuedUpdateSet.clear();
+    }
+
+    private void insertNew(Tuple_ tuple) {
+        var state = tuple.state;
+        if (state != TupleState.CREATING) {
+            propagationQueue.insert(tuple);
+        }
+    }
+
+    private void retractExisting(Tuple_ tuple) {
+        var state = tuple.state;
+        if (state.isDirty()) {
+            if (state == TupleState.DYING || state == TupleState.ABORTING) {
+                // We already retracted this tuple from another list, so we
+                // don't need to do anything
+                return;
+            }
+            propagationQueue.retract(tuple, state == TupleState.CREATING ? TupleState.ABORTING : TupleState.DYING);
+        } else {
+            propagationQueue.retract(tuple, TupleState.DYING);
+        }
+    }
+
+    private void insertIntoInnerNodeNetwork(Object toInsert) {
+        objectToOutputTuplesMap.put(toInsert, new ArrayList<>());
+        innerNodeNetwork.getRootNodesAcceptingType(toInsert.getClass())
+                .forEach(node -> ((AbstractForEachUniNode) node).insert(toInsert));
+    }
+
+    private void retractFromInnerNodeNetwork(Object toRetract) {
+        objectToOutputTuplesMap.remove(toRetract);
+        innerNodeNetwork.getRootNodesAcceptingType(toRetract.getClass())
+                .forEach(node -> ((AbstractForEachUniNode) node).retract(toRetract));
+    }
+
+    private void invalidateCache() {
+        objectToOutputTuplesMap.values().stream().flatMap(List::stream).forEach(this::retractExisting);
+        inputTupleToOutputTupleMap.clear();
+    }
+
+    private void recalculateTuples() {
+        for (var mappedTupleEntry : objectToOutputTuplesMap.entrySet()) {
+            mappedTupleEntry.getValue().clear();
+            var invalidated = mappedTupleEntry.getKey();
+            recordingTupleNode.startRecording(
```

Comment:
Arguably this API could be better. `startRecording` (`recorded`?) could return something `AutoCloseable` (perhaps even the `Recorder` instance which gets created out of sight), turning this into a try-with-resources. That away, it becomes much harder to forget to stop recording, and the `stopRecording` call doesn't even need to exist.

Reply:
Debating that, although I do know some performance optimizations are turned off inside try bodies